pallets/utxo/src/rewards.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) 2021 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://spdx.org/licenses/MIT
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author(s): C. Yap
+
+use crate::{
+    convert_to_h256, BlockAuthor, Config, Event, Pallet, RewardTotal, TransactionOutput, UtxoStore,
+    Value,
+};
+
+use frame_support::traits::Get;
+use sp_runtime::traits::{BlakeTwo256, CheckedDiv, Hash, SaturatedConversion};
+use sp_runtime::Percent;
+use sp_std::convert::TryInto;
+use sp_core::H256;
+
+/// handle event when a block author is found.
+impl<T> pallet_authorship::EventHandler<T::AccountId, T::BlockNumber> for Pallet<T>
+where
+    T: Config + pallet_authorship::Config,
+{
+    fn note_author(author: T::AccountId) {
+        match convert_to_h256::<T>(&author) {
+            Ok(author_h256) => {
+                // store the block author. Reward during the `fn finalize()` phase.
+                <BlockAuthor<T>>::put(author_h256);
+            }
+            Err(e) => {
+                log::warn!("failed to find author: {:?}", e);
+            }
+        }
+    }
+
+    fn note_uncle(_author: T::AccountId, _age: T::BlockNumber) {
+        log::info!("TODO: no support for this. Or is there...?");
+    }
+}
+
+
+/// checks at what period the given block number belongs to.
+fn get_block_period<T:Config>(block_number: T::BlockNumber) -> u8 {
+    let reduction_period: T::BlockNumber = T::RewardReductionPeriod::get();
+
+    // get at what period the current block number is at.
+    block_number.checked_div(&reduction_period)
+        .map_or(u8::MAX, |f|{
+            // convert data type from T::BlockNumber to u8
+            f.try_into().unwrap_or(u8::MAX)
+        })
+}
+
+
+/// Returns the newly reduced reward amount for a Block Author.
+/// How much a reward is reduced, will be based on the config's`RewardReductionFraction`.
+fn get_block_author_reward<T: Config>(block_number: T::BlockNumber) -> Value {
+    let reward_amount = T::InitialReward::get();
+    let reduction_period: T::BlockNumber = T::RewardReductionPeriod::get();
+    let reduction_fraction = T::RewardReductionFraction::get().deconstruct();
+
+    // no computation needed; current block number is still within the reduction period.
+    if block_number < reduction_period {
```

Comment:
This is completely unnecessary. If your math below is correct, this result will be the natural consequence of it.